File: server/src/collectors/soramame.js
/**
 * SORAMAME Air Quality Collector
 * Ministry of Environment 大気汚染物質広域監視システム ("そらまめくん").
 *
 * Tries several public SORAMAME / EnvJP endpoints (no auth required), then
 * falls back to OSM `man_made=monitoring_station` + `monitoring:air_quality=yes`
 * for any active sensor, then finally to a static seed of major stations
 * with synthetic readings (clearly marked as `soramame_seed`).
 */

import { fetchJson, fetchOverpass, fetchText } from './_liveHelpers.js';

const ATMOS_URLS = [
  // Public atmospheric stations index (lat/lon list)
  'https://soramame.env.go.jp/data/map/kyokuLatest/Pref01.json',
  'https://soramame.env.go.jp/data/jsons/japan/atmos.json',
  // Cross-government NIES backup (the AEROS feed)
  'https://tenbou.nies.go.jp/download/Air_pollution/json/StationLatest.json',
];

const STATIONS = [
  { id: 'AQ001', name: '札幌北', pref: '北海道', lat: 43.091, lon: 141.341 },
  { id: 'AQ002', name: '旭川', pref: '北海道', lat: 43.771, lon: 142.365 },
  { id: 'AQ003', name: '釧路', pref: '北海道', lat: 42.985, lon: 144.382 },
  { id: 'AQ004', name: '青森中央', pref: '青森県', lat: 40.822, lon: 140.747 },
  { id: 'AQ005', name: '盛岡', pref: '岩手県', lat: 39.702, lon: 141.153 },
  { id: 'AQ006', name: '仙台宮城野', pref: '宮城県', lat: 38.271, lon: 140.882 },
  { id: 'AQ007', name: '秋田中央', pref: '秋田県', lat: 39.720, lon: 140.103 },
  { id: 'AQ008', name: '山形', pref: '山形県', lat: 38.241, lon: 140.334 },
  { id: 'AQ009', name: '福島', pref: '福島県', lat: 37.750, lon: 140.468 },
  { id: 'AQ010', name: '水戸', pref: '茨城県', lat: 36.366, lon: 140.471 },
  { id: 'AQ011', name: '宇都宮', pref: '栃木県', lat: 36.566, lon: 139.884 },
  { id: 'AQ012', name: '前橋', pref: '群馬県', lat: 36.391, lon: 139.061 },
  { id: 'AQ013', name: 'さいたま大宮', pref: '埼玉県', lat: 35.906, lon: 139.631 },
  { id: 'AQ014', name: '千葉中央', pref: '千葉県', lat: 35.607, lon: 140.106 },
  { id: 'AQ015', name: '東京千代田', pref: '東京都', lat: 35.694, lon: 139.754 },
  { id: 'AQ016', name: '東京新宿', pref: '東京都', lat: 35.694, lon: 139.703 },
  { id: 'AQ017', name: '東京世田谷', pref: '東京都', lat: 35.646, lon: 139.653 },
  { id: 'AQ018', name: '東京大田', pref: '東京都', lat: 35.561, lon: 139.716 },
  { id: 'AQ019', name: '東京江東', pref: '東京都', lat: 35.673, lon: 139.817 },
  { id: 'AQ020', name: '横浜鶴見', pref: '神奈川県', lat: 35.510, lon: 139.682 },
  { id: 'AQ021', name: '川崎', pref: '神奈川県', lat: 35.531, lon: 139.703 },
  { id: 'AQ022', name: '新潟中央', pref: '新潟県', lat: 37.916, lon: 139.036 },
  { id: 'AQ023', name: '富山', pref: '富山県', lat: 36.695, lon: 137.211 },
  { id: 'AQ024', name: '金沢', pref: '石川県', lat: 36.594, lon: 136.626 },
  { id: 'AQ025', name: '福井', pref: '福井県', lat: 36.065, lon: 136.222 },
  { id: 'AQ026', name: '甲府', pref: '山梨県', lat: 35.664, lon: 138.568 },
  { id: 'AQ027', name: '長野', pref: '長野県', lat: 36.232, lon: 138.181 },
  { id: 'AQ028', name: '岐阜', pref: '岐阜県', lat: 35.391, lon: 136.722 },
  { id: 'AQ029', name: '静岡', pref: '静岡県', lat: 34.977, lon: 138.383 },
  { id: 'AQ030', name: '名古屋中', pref: '愛知県', lat: 35.181, lon: 136.906 },
  { id: 'AQ031', name: '名古屋南', pref: '愛知県', lat: 35.115, lon: 136.933 },
  { id: 'AQ032', name: '津', pref: '三重県', lat: 34.730, lon: 136.509 },
  { id: 'AQ033', name: '大津', pref: '滋賀県', lat: 35.005, lon: 135.869 },
  { id: 'AQ034', name: '京都中京', pref: '京都府', lat: 35.012, lon: 135.768 },
  { id: 'AQ035', name: '大阪市此花', pref: '大阪府', lat: 34.681, lon: 135.437 },
  { id: 'AQ036', name: '大阪市天王寺', pref: '大阪府', lat: 34.653, lon: 135.519 },
  { id: 'AQ037', name: '堺', pref: '大阪府', lat: 34.573, lon: 135.483 },
  { id: 'AQ038', name: '神戸中央', pref: '兵庫県', lat: 34.690, lon: 135.196 },
  { id: 'AQ039', name: '尼崎', pref: '兵庫県', lat: 34.733, lon: 135.407 },
  { id: 'AQ040', name: '奈良', pref: '奈良県', lat: 34.685, lon: 135.833 },
  { id: 'AQ041', name: '和歌山', pref: '和歌山県', lat: 34.226, lon: 135.168 },
  { id: 'AQ042', name: '岡山', pref: '岡山県', lat: 34.662, lon: 133.935 },
  { id: 'AQ043', name: '広島中区', pref: '広島県', lat: 34.396, lon: 132.460 },
  { id: 'AQ044', name: '北九州小倉', pref: '福岡県', lat: 33.883, lon: 130.875 },
  { id: 'AQ045', name: '福岡中央', pref: '福岡県', lat: 33.590, lon: 130.402 },
  { id: 'AQ046', name: '久留米', pref: '福岡県', lat: 33.319, lon: 130.508 },
  { id: 'AQ047', name: '熊本', pref: '熊本県', lat: 32.790, lon: 130.742 },
  { id: 'AQ048', name: '鹿児島', pref: '鹿児島県', lat: 31.560, lon: 130.558 },
  { id: 'AQ049', name: '那覇', pref: '沖縄県', lat: 26.335, lon: 127.681 },
  { id: 'AQ050', name: '松山', pref: '愛媛県', lat: 33.842, lon: 132.766 },
];

/**
 * Fetch each prefecture-keyed JSON ("Pref01.json" .. "Pref47.json") if the
 * primary endpoint pattern works. Returns features or null.
 */
async function tryPrefScan() {
  const features = [];
  // Iterate all 47 prefectures concurrently (low load — small JSON each)
  const tasks = [];
  for (let p = 1; p <= 47; p++) {
    const code = String(p).padStart(2, '0');
    const url = `https://soramame.env.go.jp/data/map/kyokuLatest/Pref${code}.json`;
    tasks.push(fetchJson(url, { timeoutMs: 10_000 }).then((data) => {
      if (!data || !Array.isArray(data)) return;
      for (const r of data) {
        const lat = Number(r?.緯度 ?? r?.lat ?? r?.LAT);
        const lon = Number(r?.経度 ?? r?.lon ?? r?.LON);
        const geocoded = Number.isFinite(lat) && Number.isFinite(lon);
        features.push({
          type: 'Feature',
          geometry: geocoded ? { type: 'Point', coordinates: [lon, lat] } : null,
          properties: {
            station_id: String(r?.測定局コード ?? r?.code ?? ''),
            station_name: r?.測定局名称 ?? r?.name ?? null,
            prefecture: r?.都道府県名 ?? null,
            pm25_ugm3: r?.PM25 ?? null,
            pm10_ugm3: r?.SPM ?? null,
            so2_ppb: r?.SO2 ?? null,
            no2_ppb: r?.NO2 ?? null,
            ox_ppb: r?.OX ?? null,
            co_ppm: r?.CO ?? null,
            measured_at: r?.測定時刻 ?? null,
            source: 'soramame_live',
          },
        });
      }
    }).catch(() => {}));
  }
  await Promise.all(tasks);
  return features.length > 0 ? features : null;
}

async function tryAtmosJson() {
  for (const url of ATMOS_URLS) {
    const data = await fetchJson(url, { timeoutMs: 10_000 });
    if (!data) continue;
    const arr = Array.isArray(data) ? data : (data.features || data.stations || data.list || data.data);
    if (!Array.isArray(arr) || arr.length === 0) continue;
    const features = [];
    for (const r of arr) {
      // GeoJSON feature?
      if (r?.geometry?.coordinates) {
        features.push({
          type: 'Feature',
          geometry: r.geometry,
          properties: { ...(r.properties || {}), source: 'soramame_live' },
        });
        continue;
      }
      const lat = Number(r?.lat ?? r?.LAT ?? r?.緯度);
      const lon = Number(r?.lon ?? r?.LON ?? r?.経度);
      const geocoded = Number.isFinite(lat) && Number.isFinite(lon);
      features.push({
        type: 'Feature',
        geometry: geocoded ? { type: 'Point', coordinates: [lon, lat] } : null,
        properties: {
          station_id: r?.station_id ?? r?.code ?? null,
          station_name: r?.name ?? r?.station_name ?? null,
          prefecture: r?.pref ?? null,
          pm25_ugm3: r?.pm25 ?? null,
          source: 'soramame_live',
        },
      });
    }
    if (features.length > 0) return features;
  }
  return null;
}

async function tryOsmStations() {
  return fetchOverpass(
    [
      'node["man_made"="monitoring_station"]["monitoring:air_quality"="yes"](area.jp);',
      'node["man_made"="monitoring_station"]["monitoring:weather"="yes"](area.jp);',
    ].join(''),
    (el, _i, coords) => ({
      type: 'Feature',
      geometry: { type: 'Point', coordinates: coords },
      properties: {
        station_id: `OSM_${el.id}`,
        station_name: el.tags?.['name:en'] || el.tags?.name || 'Air monitoring station',
        operator: el.tags?.operator || null,
        ref: el.tags?.ref || null,
        source: 'osm_overpass',
      },
    }),
    60_000,
    { limit: 0, queryTimeout: 120 },
  );
}

function randomInRange(min, max) {
  return Math.round((min + Math.random() * (max - min)) * 10) / 10;
}

function generateSeedData() {
  const now = new Date();
  return STATIONS.map((st) => {
    const isUrban = ['東京', '大阪', '名古屋', '横浜', '川崎', '北九州'].some(
      (c) => st.name.includes(c) || st.pref.includes(c),
    );
    const f = isUrban ? 1.4 : 1.0;
    const pm25 = randomInRange(5, 35) * f;
    return {
      type: 'Feature',
      geometry: { type: 'Point', coordinates: [st.lon, st.lat] },
      properties: {
        station_id: st.id,
        station_name: st.name,
        prefecture: st.pref,
        pm25_ugm3: Math.round(pm25 * 10) / 10,
        measured_at: now.toISOString(),
        source: 'soramame_seed',
      },
    };
  });
}

export default async function collectSoramame() {
  // Live cascade: prefecture scan → atmos JSON → OSM monitoring stations → seed
  let features = await tryPrefScan();
  let live = !!(features && features.length);
  let liveSrc = live ? 'soramame_pref_scan' : null;

  if (!live) {
    features = await tryAtmosJson();
    live = !!(features && features.length);
    if (live) liveSrc = 'soramame_atmos';
  }
  if (!live) {
    features = await tryOsmStations();
    live = !!(features && features.length);
    if (live) liveSrc = 'osm_monitoring_station';
  }
  if (!live) features = [];

  return {
    type: 'FeatureCollection',
    features: features || [],
    _meta: {
      source: live ? liveSrc : 'soramame_seed',
      fetchedAt: new Date().toISOString(),
      recordCount: features?.length || 0,
      live,
      description: 'SORAMAME / NIES air quality monitoring stations across Japan',
    },
  };
}

File: native/collectors/sources/soramame.c
/* collectors/environment/sources/soramame.c
 * Port of server/src/collectors/soramame.js.
 * Live cascade (rule 7 — generateSeedData not ported):
 *   1. tryPrefScan: Pref01..Pref47.json (keyless)
 *   2. tryAtmosJson: ATMOS_URLS list
 *   3. tryOsmStations: OSM Overpass monitoring_station
 * First non-empty wins. */
#include "source.h"
#include "lib/feedlib.h"
#include "lib/geojson.h"
#include "lib/overpass.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* Number(r?.a ?? r?.b ?? r?.c) → finite double, or NAN. */
static int num_of(cJSON *r, const char *a, const char *b, const char *c,
                  double *out) {
  const char *ks[3] = { a, b, c };
  for (int i = 0; i < 3; i++) {
    cJSON *v = cJSON_GetObjectItem(r, ks[i]);
    if (!v) continue;
    if (cJSON_IsNumber(v)) { *out = v->valuedouble; return 1; }
    if (cJSON_IsString(v) && v->valuestring && v->valuestring[0]) {
      char *end; double d = strtod(v->valuestring, &end);
      if (end != v->valuestring) { *out = d; return 1; }
    }
  }
  return 0;
}

/* r?.x ?? r?.y → cJSON value duplicate, or JSON null. */
static cJSON *pick2(cJSON *r, const char *x, const char *y) {
  cJSON *v = cJSON_GetObjectItem(r, x);
  if (v && !cJSON_IsNull(v)) return cJSON_Duplicate(v, 1);
  v = cJSON_GetObjectItem(r, y);
  if (v && !cJSON_IsNull(v)) return cJSON_Duplicate(v, 1);
  return cJSON_CreateNull();
}
static cJSON *pick1(cJSON *r, const char *x) {
  cJSON *v = cJSON_GetObjectItem(r, x);
  return (v && !cJSON_IsNull(v)) ? cJSON_Duplicate(v, 1) : cJSON_CreateNull();
}

/* String(r?.測定局コード ?? r?.code ?? '') */
static void str_id(cJSON *r, const char *a, const char *b, char *out,
                   size_t n) {
  cJSON *v = cJSON_GetObjectItem(r, a);
  if (!(v && !cJSON_IsNull(v))) v = cJSON_GetObjectItem(r, b);
  if (v && cJSON_IsString(v) && v->valuestring) {
    snprintf(out, n, "%s", v->valuestring);
  } else if (v && cJSON_IsNumber(v)) {
    snprintf(out, n, "%lld", (long long)v->valuedouble);
  } else {
    out[0] = '\0';
  }
}

/* Build the pref-scan Feature (JS prop order). */
static cJSON *pref_feature(cJSON *r) {
  double lat, lon;
  int geo = num_of(r, "緯度", "lat", "LAT", &lat) &&
            num_of(r, "経度", "lon", "LON", &lon);
  cJSON *f = cJSON_CreateObject();
  cJSON_AddStringToObject(f, "type", "Feature");
  if (geo) {
    cJSON *g = cJSON_CreateObject();
    cJSON_AddStringToObject(g, "type", "Point");
    cJSON *c = cJSON_CreateArray();
    cJSON_AddItemToArray(c, cJSON_CreateNumber(lon));
    cJSON_AddItemToArray(c, cJSON_CreateNumber(lat));
    cJSON_AddItemToObject(g, "coordinates", c);
    cJSON_AddItemToObject(f, "geometry", g);
  } else {
    cJSON_AddItemToObject(f, "geometry", cJSON_CreateNull());
  }
  cJSON *p = cJSON_CreateObject();
  char sid[64];
  str_id(r, "測定局コード", "code", sid, sizeof sid);
  cJSON_AddStringToObject(p, "station_id", sid);
  cJSON_AddItemToObject(p, "station_name", pick2(r, "測定局名称", "name"));
  cJSON_AddItemToObject(p, "prefecture", pick1(r, "都道府県名"));
  cJSON_AddItemToObject(p, "pm25_ugm3", pick1(r, "PM25"));
  cJSON_AddItemToObject(p, "pm10_ugm3", pick1(r, "SPM"));
  cJSON_AddItemToObject(p, "so2_ppb", pick1(r, "SO2"));
  cJSON_AddItemToObject(p, "no2_ppb", pick1(r, "NO2"));
  cJSON_AddItemToObject(p, "ox_ppb", pick1(r, "OX"));
  cJSON_AddItemToObject(p, "co_ppm", pick1(r, "CO"));
  cJSON_AddItemToObject(p, "measured_at", pick1(r, "測定時刻"));
  cJSON_AddStringToObject(p, "source", "soramame_live");
  cJSON_AddItemToObject(f, "properties", p);
  return f;
}

static cJSON *atmos_feature(cJSON *r) {
  cJSON *geom = cJSON_GetObjectItem(r, "geometry");
  cJSON *gcoords = geom ? cJSON_GetObjectItem(geom, "coordinates") : NULL;
  cJSON *f = cJSON_CreateObject();
  cJSON_AddStringToObject(f, "type", "Feature");
  if (gcoords) {
    cJSON_AddItemToObject(f, "geometry", cJSON_Duplicate(geom, 1));
    cJSON *p = cJSON_CreateObject();
    cJSON *props = cJSON_GetObjectItem(r, "properties");
    if (cJSON_IsObject(props)) {
      cJSON *kv;
      cJSON_ArrayForEach(kv, props) {
        if (kv->string && !strcmp(kv->string, "source")) continue;
        cJSON_AddItemToObject(p, kv->string, cJSON_Duplicate(kv, 1));
      }
    }
    cJSON_AddStringToObject(p, "source", "soramame_live");
    cJSON_AddItemToObject(f, "properties", p);
    return f;
  }
  double lat, lon;
  int geo = num_of(r, "lat", "LAT", "緯度", &lat) &&
            num_of(r, "lon", "LON", "経度", &lon);
  if (geo) {
    cJSON *g = cJSON_CreateObject();
    cJSON_AddStringToObject(g, "type", "Point");
    cJSON *c = cJSON_CreateArray();
    cJSON_AddItemToArray(c, cJSON_CreateNumber(lon));
    cJSON_AddItemToArray(c, cJSON_CreateNumber(lat));
    cJSON_AddItemToObject(g, "coordinates", c);
    cJSON_AddItemToObject(f, "geometry", g);
  } else {
    cJSON_AddItemToObject(f, "geometry", cJSON_CreateNull());
  }
  cJSON *p = cJSON_CreateObject();
  cJSON_AddItemToObject(p, "station_id", pick2(r, "station_id", "code"));
  cJSON_AddItemToObject(p, "station_name", pick2(r, "name", "station_name"));
  cJSON_AddItemToObject(p, "prefecture", pick1(r, "pref"));
  cJSON_AddItemToObject(p, "pm25_ugm3", pick1(r, "pm25"));
  cJSON_AddStringToObject(p, "source", "soramame_live");
  cJSON_AddItemToObject(f, "properties", p);
  return f;
}

static cJSON *osm_map(cJSON *el, int i, double lon, double lat, void *ud) {
  (void)i; (void)ud;
  cJSON *f = gj_point_feature(lon, lat);
  cJSON *p = cJSON_CreateObject();
  cJSON *id = cJSON_GetObjectItem(el, "id");
  char sid[48];
  snprintf(sid, sizeof sid, "OSM_%lld",
           id && cJSON_IsNumber(id) ? (long long)id->valuedouble : 0);
  cJSON_AddStringToObject(p, "station_id", sid);
  const char *nm = ov_tag(el, "name:en");
  if (!nm) nm = ov_tag(el, "name");
  cJSON_AddStringToObject(p, "station_name",
                          nm ? nm : "Air monitoring station");
  const char *op = ov_tag(el, "operator");
  if (op) cJSON_AddStringToObject(p, "operator", op);
  else cJSON_AddItemToObject(p, "operator", cJSON_CreateNull());
  const char *ref = ov_tag(el, "ref");
  if (ref) cJSON_AddStringToObject(p, "ref", ref);
  else cJSON_AddItemToObject(p, "ref", cJSON_CreateNull());
  cJSON_AddStringToObject(p, "source", "osm_overpass");
  cJSON_AddItemToObject(f, "properties", p);
  return f;
}

static const char *ATMOS_URLS[] = {
  "https://soramame.env.go.jp/data/map/kyokuLatest/Pref01.json",
  "https://soramame.env.go.jp/data/jsons/japan/atmos.json",
  "https://tenbou.nies.go.jp/download/Air_pollution/json/StationLatest.json",
  NULL
};

static int run(const source_ctx *ctx, intel_sink *sink) {
  /* 1. tryPrefScan: Pref01..Pref47.json */
  cJSON *features = cJSON_CreateArray();
  for (int pi = 1; pi <= 47; pi++) {
    char url[160];
    snprintf(url, sizeof url,
      "https://soramame.env.go.jp/data/map/kyokuLatest/Pref%02d.json", pi);
    cJSON *data = feed_get_json(ctx->http, url, 10000);
    if (data && cJSON_IsArray(data)) {
      cJSON *r;
      cJSON_ArrayForEach(r, data)
        cJSON_AddItemToArray(features, pref_feature(r));
    }
    if (data) cJSON_Delete(data);
  }
  if (cJSON_GetArraySize(features) > 0) {
    int n = geojson_emit_features(sink, ctx->source_id, features);
    cJSON_Delete(features);
    fprintf(stderr, "[soramame] pref-scan emitted %d\n", n);
    return n >= 0 ? 0 : -1;
  }
  cJSON_Delete(features);

  /* 2. tryAtmosJson */
  for (int u = 0; ATMOS_URLS[u]; u++) {
    cJSON *data = feed_get_json(ctx->http, ATMOS_URLS[u], 10000);
    if (!data) continue;
    cJSON *arr = NULL;
    if (cJSON_IsArray(data)) arr = data;
    else {
      const char *keys[] = { "features", "stations", "list", "data" };
      for (int k = 0; k < 4 && !arr; k++) {
        cJSON *a = cJSON_GetObjectItem(data, keys[k]);
        if (cJSON_IsArray(a)) arr = a;
      }
    }
    if (!cJSON_IsArray(arr) || cJSON_GetArraySize(arr) == 0) {
      cJSON_Delete(data);
      continue;
    }
    cJSON *feats = cJSON_CreateArray();
    cJSON *r;
    cJSON_ArrayForEach(r, arr)
      cJSON_AddItemToArray(feats, atmos_feature(r));
    cJSON_Delete(data);
    if (cJSON_GetArraySize(feats) > 0) {
      int n = geojson_emit_features(sink, ctx->source_id, feats);
      cJSON_Delete(feats);
      fprintf(stderr, "[soramame] atmos emitted %d\n", n);
      return n >= 0 ? 0 : -1;
    }
    cJSON_Delete(feats);
  }

  /* 3. tryOsmStations */
  int n = overpass_collect(ctx, sink,
    "node[\"man_made\"=\"monitoring_station\"][\"monitoring:air_quality\"=\"yes\"](area.jp);"
    "node[\"man_made\"=\"monitoring_station\"][\"monitoring:weather\"=\"yes\"](area.jp);",
    120, 60000, osm_map, NULL);
  return n >= 0 ? 0 : -1;
}

static const source_def soramame_def = {
  .id = "soramame", .collector = "environment",
  .name = "SORAMAME Air Quality", .name_ja = "そらまめ君 大気汚染情報",
   .update_interval_sec = 3600, .run = run };
REGISTER_SOURCE(soramame_def)
